Fixed default buffer arguments with binding [Fixes `bufaccess`]

da-woods opened the pull request:
They were assigned to global scope, but buffers aren't allowed in
global scope. To fix, dropped the buffer type from the default value.

--------

Fixes the compile error. However a bunch of reference counting "errors" appear (i.e. numbers don't match existing numbers). Link to https://github.com/cython/cython/pull/2864

scoder commented:
Hmm … maybe it's just an unsupported feature then? It's probably ok to document that and require users to mark these functions with `@binding(False)` (also in our tests).

da-woods commented:
The reference counting errors are a non-issue - it's the same as the other reference counting with `binding=True` - the reference count is usually +1 for some reason. It's only that the unit tests were out. I'm relatively happy that this patch is valid - it just doesn't deal with the unit tests.

da-woods commented:
I've updated the reference counts so the tests now pass fully. I've also updated the title since it wasn't obvious which failure this (and my other commit) was actually fixing which was probably unhelpful.

I believe the extra reference (also for the tests @mattip updated a little while back) comes from https://github.com/cython/cython/blob/c75ed142e54351596a6de2359295730c5fab5ec6/Cython/Utility/CythonFunction.c#L679. There's a reference for the original object, a reference to store it in `args` and a reference to store it in `new_args` hence the "base reference count" is 3. For the tests in this file I've decided that the changes are more important than the absolute values so tested those instead. However, if you want to see absolute values then I can put them back.

mattip commented:
ping

scoder commented:
Hmm, sorry, I guess I shouldn't have rebased the target branch but merged it (will do that next time). Would you mind rebasing (or cherry-picking) this PR so that we get a clean diff again?

da-woods commented:
Re-based

scoder commented:
Thanks
